### PR TITLE
Private events rails project covering Advanced Associations and Active Record Queries

### PR DESCRIPTION
**Pull request by KabohaJeanMark**
- Create a professional Microverse Repo
- Add a descriptive README with information about the Project for understanding and set up guidance.
- Set up and use devise for authentication.
- Use the has _many associations for creators and the events
- Use Advanced Active Record Associations to model and map the has_many relationship using a through table to deal with the many to many relationship between events and attendees.
- Use Active Record query methods and scope to display attended events as per the order of upcoming and past.
- Add page links to the application.html.erb file only showing creation and user-created events links for logged in users

### CODE REVIEW

File: app/views/users/show.html.erb [hunk 1]
```diff
@@ -0,0 +1,29 @@
+<h1>List of Events you are attending</h1>
+
+<h2>Upcoming events</h2>
+<ul>
+  <% current_user.attended_events.where('date >= ?', DateTime.now).each do |event| %>
```
Comment (bini-i) [inline app/views/users/show.html.erb]: Kindly use the scope from Event model and pass it as an instance variable to your views, as it helps you to keep your views clean from logic and presentation only.
`@upcoming_events = current_user.attended_events.upcoming`

File: app/views/users/show.html.erb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    <li>
+      <%= event.name%>
+    </li>
+  <% end %>
+</ul>
+
+<h2>Past events</h2>
+<ul>
+  <% current_user.attended_events.where('date < ?', DateTime.now).each do |event| %>
```
Comment (bini-i) [inline app/views/users/show.html.erb]: Kindly use the scope from Event model and pass it as an instance variable to your views, as it helps you to keep your views clean from logic and presentation only.
`@past_events = current_user.attended_events.previous`

**Review by bini-i — CHANGES_REQUESTED** [review]
# Status: Changes Required :yellow_circle:            
### Hello @KabohaJeanMark  :wave: 

You have done very good :clap: for a first submission, the following are some of the highlights of your project:
 - descriptive and well written README and pull request description

Please work on the following missing requirements to further enhance your project:
#### To improve
- Navigation-link to the root directory is missing at the top of the page, kindly add one to make it easier to jump around between pages.
- Even though your model shows the one-to-many relationships between a creator and an event, in the database the event table is missing a foreign key that references to the user table as a creator of the event. Please add a foreign key to the events table that makes a reference to the user table.
- see inline comments above :point_up:  

#### Suggestions
- Since you have a good foundation of HTML and CSS by now, it would help your project standout if you add styling to it.

#### If you have any question feel free to ask within the PR and ping me on slack @Binyam Hailemeskel

### HAPPY CODING :computer: :blush: 

> #### Reviewed by Binyam Hailemeskel (Code Reviewer)
> GitHub: [@bini-i](https://github.com/bini-i) 
> Twitter: [@binyamshewa](https://twitter.com/binyamshewa)
> LinkedIn: [LinkedIn](https://www.linkedin.com/in/binyam-hailemeskel-728048151/)#

File: app/views/users/show.html.erb (hunk 1 above)
Comment (KabohaJeanMark) [reply app/views/users/show.html.erb]: Thanks for the clear precise feedback.
I follow your rationale and have implemented the changes using the upcoming method from the Event model scope.

File: app/views/users/show.html.erb (hunk 2 above)
Comment (KabohaJeanMark) [reply app/views/users/show.html.erb]: Thanks. I have implemented this change similar to the one above by using the Event scope past method.

**Comment by KabohaJeanMark** [comment]
Thank you 👏🏾 for the clear actionable, specific and kind feedback @bini-i.
I agree with what you have pointed out and it helps improve the project's clarity and visibility by properly utilising the scope to reduce the logic in the views for better separation of concerns following MVC design patterns best principles.
Actually, the User id is present in the events table as the foreign key in a column called creator_id. 
I have also added some bit of styling to the project to work on all mandatory and optional changes 💪🏾

**Review by george-shammar — APPROVED** [review]
### **STATUS: APPROVED** :heavy_check_mark: :heavy_check_mark: :+1: 

Hello, good job for implementing the required changes. :+1: 
**Highlights**

- [ ] You now have a navigation bar that makes jumping between pages possible.
- [ ] You have a foreign key in your events table pointing to the user table.

Well done :+1: 
[OPTIONAL]

- [ ] I strongly advise you to include tests in your project. Testing is an important requirement in this project and it prepares you for the more complicated projects ahead especially the ror/social_media app. Add unit tests for model associations and validations (Use Rspec). Your authentication and event pages should also have integration tests. (Use Rspec and Capybara). I'm making this optional because you've already made the required changes from the previous review.

I wish you good luck with your future projects.

Cheers :smile:

**Pull request merged**